### PR TITLE
perf: improve compress impl

### PR DESCRIPTION
**Pull request by driftluo**
### What problem does this PR solve?

Optimize message compression and incorporate it into the codec to reduce the consumption of a full message copy for each communication

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

**Review by eval-exec — DISMISSED** [review]

**Review by eval-exec — DISMISSED** [review]

File: network/src/compress.rs [hunk 1]
```diff
@@ -116,3 +117,142 @@ pub fn compress(src: Bytes) -> Bytes {
...
+                            if decompressed_bytes_len > MAX_UNCOMPRESSED_LEN {
+                                debug!(
+                                    "The limit for uncompressed bytes len is exceeded. limit: {}, len: {}",
+                                    MAX_UNCOMPRESSED_LEN, decompressed_bytes_len
+                                );
+                                return Err(io::ErrorKind::InvalidData.into());
+                            }
+                            let mut buf = vec![0; decompressed_bytes_len];
+                            match SnapDecoder::new().decompress(&data[1..], &mut buf) {
+                                Ok(_) => Ok(Some(BytesMut::from(Bytes::from(buf)))),
```
Comment (quake) [inline network/src/compress.rs]: here is a data copy when convert buf to Bytes, can we use decompress_vec to write the data to BytesMut directly?

Comment (driftluo) [reply network/src/compress.rs]: > Convert self into BytesMut.

> If bytes is unique for the entire original buffer, this will return a BytesMut with the contents of bytes without copying. If bytes is not unique for the entire original buffer, this will make a copy of bytes subset of the original buffer in a new BytesMut.

From the code implementation perspective, there is no copying, but relying on the internal implementation behavior is not good. I will change to：
```rust
let mut buf = BytesMut::zeroed(decompressed_bytes_len);
match SnapDecoder::new().decompress(&data[1..], &mut buf) {
    Ok(_) => Ok(Some(buf)),
    Err(e) => {
        debug!("snappy decompress error: {:?}", e);
        Err(io::ErrorKind::InvalidData.into())
    }
}
```

**Review by eval-exec — APPROVED** [review]

File: network/src/compress.rs [hunk 2]
```diff
@@ -116,3 +117,142 @@ pub fn compress(src: Bytes) -> Bytes {
...
+            enable_compress,
+            protocol_id,
+        }
+    }
+}
+
+impl tokio_util::codec::Encoder<Bytes> for LengthDelimitedCodecWithCompress {
+    type Error = io::Error;
+    fn encode(&mut self, data: Bytes, dst: &mut BytesMut) -> Result<(), io::Error> {
+        let process = |data: &[u8], flag: u8, dst: &mut BytesMut| {
```
Comment (quake) [inline network/src/compress.rs]: Defining this logic as a helper method instead of an inline closure would make the code easier to read and reuse, and keeps the encode function more concise.

```suggestion
impl LengthDelimitedCodecWithCompress {
    fn process(&self, data: &[u8], flag: u8, dst: &mut BytesMut) -> { ... }
}
```

File: network/src/compress.rs [hunk 3]
```diff
@@ -116,3 +117,142 @@ pub fn compress(src: Bytes) -> Bytes {
...
+impl tokio_util::codec::Decoder for LengthDelimitedCodecWithCompress {
+    type Item = BytesMut;
+    type Error = io::Error;
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        if src.is_empty() {
+            return Ok(None);
+        }
+        match self.length_delimited.decode(src)? {
+            Some(mut data) => {
+                if (data[0] & COMPRESS_FLAG) != 0 {
```
Comment (quake) [inline network/src/compress.rs:232]: Accessing data[0] and data[1..] can panic if data is empty. A malicious peer could send a crafted frame with no payload (for example just the 4-byte length header like [0, 0, 0, 0]), need to check data length first. Also, please add a corresponding unit test to cover this edge case and verify that malformed or empty frames are handled safely without crashing.

```rust
    let mut buf = BytesMut::from(vec![0, 0, 0, 0]);
    let decoded = codec_with_compress.decode(&mut buf);
```

**Review by quake — CHANGES_REQUESTED** [review]

File: network/src/compress.rs [hunk 4]
```diff
@@ -116,3 +117,147 @@ pub fn compress(src: Bytes) -> Bytes {
...
+        protocol_id: p2p::ProtocolId,
+    ) -> Self {
+        Self {
+            length_delimited,
+            enable_compress,
+            protocol_id,
+        }
+    }
+
+    fn process(&mut self, data: &[u8], flag: u8, dst: &mut BytesMut) -> Result<(), io::Error> {
```
Comment (quake) [inline network/src/compress.rs]:
```suggestion
    fn process(&self, data: &[u8], flag: u8, dst: &mut BytesMut) -> Result<(), io::Error> {
```

**Review by quake — APPROVED** [review]
